homeassistant/components/rainmachine/switch.py

```diff
@@ -237,171 +236,211 @@ async def _async_run_switch_coroutine(self, api_coro: Coroutine) -> None:
             async_update_programs_and_zones(self.hass, self._entry)
         )
 
-    async def async_disable_program(self) -> None:
-        """Disable a program."""
-        raise NotImplementedError("Service not implemented for this entity")
-
-    async def async_disable_zone(self) -> None:
-        """Disable a zone."""
-        raise NotImplementedError("Service not implemented for this entity")
-
-    async def async_enable_program(self) -> None:
-        """Enable a program."""
-        raise NotImplementedError("Service not implemented for this entity")
-
-    async def async_enable_zone(self) -> None:
-        """Enable a zone."""
-        raise NotImplementedError("Service not implemented for this entity")
-
     async def async_start_program(self) -> None:
-        """Start a program."""
+        """Execute the start_program entity service."""
         raise NotImplementedError("Service not implemented for this entity")
 
     async def async_start_zone(self, *, zone_run_time: int) -> None:
-        """Start a zone."""
+        """Execute the start_zone entity service."""
         raise NotImplementedError("Service not implemented for this entity")
 
     async def async_stop_program(self) -> None:
-        """Stop a program."""
+        """Execute the stop_program entity service."""
         raise NotImplementedError("Service not implemented for this entity")
 
     async def async_stop_zone(self) -> None:
-        """Stop a zone."""
+        """Execute the stop_zone entity service."""
         raise NotImplementedError("Service not implemented for this entity")
 
-    @callback
-    def update_from_latest_data(self) -> None:
-        """Update the state."""
-        self._data = self.coordinator.data[self.entity_description.uid]
-        self._is_active = self._data["active"]
 
+class RainMachineActivitySwitch(RainMachineBaseSwitch):
+    """Define a RainMachine switch to start/stop an activity (program or zone)."""
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the switch off.
+
+        The only way this could occur is if someone rapidly turns a disabled activity
+        off right after turning it on.
+        """
+        if not self.coordinator.data[self.entity_description.uid]["active"]:
+            self._attr_is_on = False
+            self.async_write_ha_state()
+            return
+
+        await self.async_turn_off_when_active(**kwargs)
+
+    async def async_turn_off_when_active(self, **kwargs: Any) -> None:
+        """Turn the switch off when its associated activity is active."""
+        raise NotImplementedError
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn the switch on."""
+        if not self.coordinator.data[self.entity_description.uid]["active"]:
+            LOGGER.error("Cannot turn on an inactive program/zone: %s", self.name)
```

Comment:
We should raise an exception with a base class of `HomeAssistantError` so the user gets feedback about why the zone can't be turned on in the UI via a toast message.